src/object/cli_obj.c

```diff
@@ -139,9 +139,6 @@ obj_shard_open(struct dc_object *obj, unsigned int shard, unsigned int map_ver,
 		D_GOTO(unlock, rc = -DER_NONEXIST);
 	}
 
-	/* XXX could be otherwise for some object classes? */
-	D_ASSERT(obj_shard->do_shard == shard);
```

Comment:
In what use case they are not equal?

Reply:
In this patches scenario.

When performing online features (reintegration/drain/etc.) we need to return an extended layout containing the shards we're moving data to, that way when we perform concurrent I/O the new writes also go to the targets we're reintegrating.

In this case we need a way keep track of which shards are in the extended layout by giving them the same shard number.

For example
```
[T1 T2 T3] <- original layout.
[S1 S2 S3] <- shard ordering

[T4 T2 T3] <- new layout after failure and rebuild.
[S1 S2 S3]

[T4 T2 T3 T1] <- for online reintegration, we need to write to everything because of concurrent I/O.
[S1 S2 S3 S1] <- Reintegrating target being reintegrated for shard 1

[T1 T2 T3] <- reintegration complete.
[S1 S2 S3]
```